### PR TITLE
docs(:lock:): add docs to explain how UserSession works

### PR DESCRIPTION
**Pull request by jgravois**
ISSUES CLOSED: #224

first pass at some conceptual documentation that explains how to instantiate and work with a `UserSession` to access secure content.

adds _yet another_ browser demo to show how `ArcGISAuthError.retry()` and OAuth 2.0 fit together _without_ fun stuff like vue or pulling a serialized session from a cookie.

### CODE REVIEW

File: docs/src/guides/node.md [hunk 1]
```diff
@@ -26,5 +26,43 @@ request("https://www.arcgis.com/sharing/rest/info").then(response => {
   console.log(response);
 });
 ```
+Demo - [express](https://github.com/Esri/arcgis-rest-js/tree/master/demos/express)
+
+## User Authentication
```
Comment (tomwayson) [inline docs/src/guides/node.md]: Wouldn't it be better to show [app auth](https://developers.arcgis.com/rest/users-groups-and-items/authentication.htm#GUID-2BBC575E-F867-42F4-96F4-DA4DA678F6CC) for node? Isn't that going to be more of the norm?

Or maybe show both?

File: docs/src/guides/browser-authentication.md [hunk 2]
```diff
@@ -7,3 +7,58 @@ group: 2-authentication
 ---
 
 # Authentication in Browser-based Apps
+
+In the [Node.js](../node/) guide we explained how to instantiate a [`UserSession`](/arcgis-rest-js/api/auth/UserSession/) with hardcoded credentials.
+
+```js
```
Comment (tomwayson) [inline docs/src/guides/browser-authentication.md]: I wouldn't even put this snippet in this page.

File: demos/oauth2-browser-retry/index.html [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      });
+
+      function appendSession () {
+        // Begin an OAuth2 login using a popup.
+        return arcgisRest.UserSession.beginOAuth2({
+          clientId: clientId.defaultValue,
+          redirectUri: redirect_uri + 'authenticate.html',
+          popup: true
+        })
+          .then(function (session) {
```
Comment (tomwayson) [inline demos/oauth2-browser-retry/index.html]: is this `.then()` block even needed?

File: docs/src/guides/browser-authentication.md [hunk 4]
```diff
@@ -7,3 +7,58 @@ group: 2-authentication
...
+        .then(response => console.log(response))
+    }
+  })
+
+appendSession = () => {
+  return UserSession.beginOAuth2({
+    clientId,
+    redirectUri: 'https://yourapp.com/authenticate.html'
+  })
+  .then(session) {
```
Comment (tomwayson) [inline docs/src/guides/browser-authentication.md]: is this `.then()` block even needed?

File: docs/src/guides/browser-authentication.md [hunk 5]
```diff
@@ -7,3 +7,58 @@ group: 2-authentication
...
+
+![browser based login](https://developers.arcgis.com/documentation/core-concepts/security-and-authentication/images/authorization-screen.png)
+
+After they've signed in the `session` will keep track of individual `trustedServers` that are known to be federated.
+
+Demo - [OAuth 2.0 Browser](https://github.com/Esri/arcgis-rest-js/tree/master/demos/oauth2-browser)
+
+## Retrying requests
+
+If you don't _know_ if a web request requires authentication, you can try it anonymously and call `ArcGISAuthError.retry()` if you encounter a `403/499`.
```
Comment (tomwayson) [inline docs/src/guides/browser-authentication.md]: It might be best to move this whole description over to the README for this demo and just link there?

**Comment by jgravois** [comment]
pushed a few more commits to:

* explain when to use App Login and when to use a `UserSession` in a node.js application.
* tweak the batch-geocoder demo to follow best practice
* move info about `retry()` into the demo README
* remove the unnecessary code @tomwayson pointed out in my snippets.

**Pull request merged**